### PR TITLE
Solidity Tip and AirDrop tested 

### PR DESCRIPTION
**Pull request by earljohn004**

### CODE REVIEW

File: solidity/contracts/TipBot.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      msg.value 
+    );
+  }
+
+
+  function airDrop ( 
+    address payable[] memory accountAddress
+  ) public payable nonReentrant {
+
+    uint256 distributedAmount = msg.value / accountAddress.length;
```
Comment (osiastedian) [inline solidity/contracts/TipBot.sol]: This should be 

```solidity
((1 ether - airdropRate) * msg.value)/1 ether/accountAddressess.length
```

File: solidity/contracts/TipBot.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+      emit Tip( 
+        msg.sender,
+        accountAddress[i],
+        msg.value 
+      );
+    }
+  }
+
+  function Withdraw (
```
Comment (osiastedian) [inline solidity/contracts/TipBot.sol]: This should only be camelCase. And events should be TitleCase.

```suggestion
  function withdraw (
```

File: solidity/flatten/TipBot.sol [hunk 3]
```diff
@@ -0,0 +1,2129 @@
+// Dependency file: @openzeppelin/contracts/access/IAccessControl.sol
```
Comment (osiastedian) [inline solidity/flatten/TipBot.sol]: No need to commit this.

File: solidity/package.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  "name": "TipBot",
+  "version": "1.0.0",
+  "description": "TipBot smart contract",
+  "main": "index.js",
+  "license": "MIT",
+  "scripts": {
+    "build": "truffle compile",
+    "deploy": "truffle migrate --network development",
+    "testnet": "truffle migrate --network tanenbaum",
+    "mainnet": "truffle migrate --network tanenbaum",
```
Comment (osiastedian) [inline solidity/package.json:12]: setup truffle config for `mainnet`

File: solidity/test/tip-bot.test.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    it("Should be able change airDrop rate not greater than 1 eth", async() => {
+        await tipbot.setAirdropRate( web3.utils.toWei('0.4'));
+        const feeRate = await tipbot.airdropRate();
+        const feeRateEth = await web3.utils.fromWei(feeRate);
+        expect(feeRateEth).to.be.equal('0.4');
+    });
+
+    it("Should be able to tip user1 to user2", async () => {
+
+        var user1_initial_balance = await web3.eth.getBalance(user1);
```
Comment (osiastedian) [inline solidity/test/tip-bot.test.js]: use `const` for one-item variable assignments otherwise use`let`

File: tsconfig.json [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
-    "baseUrl": ".",
-    "paths": {
-      "@db/*": ["db/*"],
-      "@db": ["db/index"]
-    }
-  },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
-  "exclude": ["node_modules"]
-}
+    "compilerOptions": {
```
Comment (osiastedian) [inline tsconfig.json]: Remove all changes for this file.

**Review by osiastedian — APPROVED** [review]

**Review by osiastedian — APPROVED** [review]

File: solidity/contracts/TipBot.sol (hunk 1 above)
Comment (earljohn004) [reply solidity/contracts/TipBot.sol]: Just a confirmation @osiastedian , 
Line number 63 already contains the logic for deducting the percentage for the distributed Amount. 
```javascript
uint256 newBalance = ((1 ether - airdropRate) * distributedAmount)/1 ether;
```

If I modify this part, our contract will now get 0.1% from 
1. the total amount to be airdrop, and 
2. from the distributed to each address

File: solidity/contracts/TipBot.sol (hunk 2 above)
Comment (earljohn004) [reply solidity/contracts/TipBot.sol]: Modified

File: solidity/flatten/TipBot.sol (hunk 3 above)
Comment (earljohn004) [reply solidity/flatten/TipBot.sol]: Modified

File: solidity/package.json (hunk 4 above)
Comment (earljohn004) [reply solidity/package.json:12]: Added truffle main net in config

File: solidity/test/tip-bot.test.js (hunk 5 above)
Comment (earljohn004) [reply solidity/test/tip-bot.test.js]: Modified using `let`

File: tsconfig.json (hunk 6 above)
Comment (earljohn004) [reply tsconfig.json]: Reverted to commited tsconfig

File: solidity/contracts/TipBot.sol (hunk 1 above)
Comment (earljohn004) [reply solidity/contracts/TipBot.sol]: I added parenthesis for clearer computation. 
```javascript
uint256 distributedAmount = ( ((1 ether - airdropRate) * msg.value ) / 1 ether ) / accountAddress.length;
```

**Pull request merged**